src/StochasticNoiseSimulator.cpp

```diff
@@ -248,19 +256,47 @@ void StochasticNoiseSimulator::runStochSimulationForId(unsigned int
                     targets  = classic_op->getOperation()->getTargets();
                     controls = classic_op->getOperation()->getControls();
                     if (!execute_op) {
-                        // applyNoiseOperation(targets.front(), controls, identity_DD, (std::unique_ptr<dd::Package> &) localDD, localRootEdge, generator, dist, line, identity_DD);
+                        // applyNoiseOperation(targets.front(), controls, identityDD, (std::unique_ptr<dd::Package> &) localDD, localRootEdge, generator, dist, line, identityDD);
                         continue;
                     }
                 } else {
-                    dd_op    = dd::getDD(op.get(), localDD);
                     targets  = op->getTargets();
                     controls = op->getControls();
+
+                    if (targets.size() == 1 && controls.empty()) {
+                        dd_op = localDD->stochasticNoiseOperationCache.lookup(op->getType(), targets.front());
+                        if (dd_op.p == nullptr) {
+                            dd_op = dd::getDD(op.get(), localDD);
+                            localDD->stochasticNoiseOperationCache.insert(op->getType(), targets.front(), dd_op);
+                        }
+                    } else {
+                        dd_op = dd::getDD(op.get(), localDD);
+                    }
+                }
+
+                std::vector usedQubits = targets;
+                for (auto control: controls) {
+                    usedQubits.push_back(control.qubit);
+                }
+
+                if (sequentialApplyNoise) {
+                    auto tmp0 = localDD->multiply(dd_op, localRootEdge);
+                    localDD->incRef(tmp0);
+                    localDD->decRef(localRootEdge);
+                    localRootEdge                    = tmp0;
+                    dd_op                            = identityDD;
+                    std::string tmp_gate_noise_types = "0";
+                    for (char gate_noise_type: gateNoiseEffects) {
+                        tmp_gate_noise_types[0] = gate_noise_type;
+                        applyNoiseOperation(usedQubits, dd_op, localDD, localRootEdge, generator, dist, identityDD, tmp_gate_noise_types);
+                    }
+                } else {
+                    applyNoiseOperation(usedQubits, dd_op, localDD, localRootEdge, generator, dist, identityDD, gateNoiseEffects);
                 }
```

Comment:
this can be encapsulated into a function (which probably resides in the DD package)

Reply:
As discussed. Moving it into the DD Package causes visibility problems. The entire stochastic noise functionality is moved to the QFR